planner/cascades/transformation_rules.go

```diff
@@ -237,6 +243,88 @@ func (r *EnumeratePaths) OnTransform(old *memo.ExprIter) (newExprs []*memo.Group
 	return newExprs, true, false, nil
 }
 
+// PushAggDownGather splits Aggregation to two stages, final and partial1,
+// and pushed the partial Aggregation down to the child of TableGather.
+type PushAggDownGather struct {
+}
+
+// GetPattern implements Transformation interface. The pattern of this rule
+// is `Aggregation -> TableGather`.
+func (r *PushAggDownGather) GetPattern() *memo.Pattern {
+	return memo.BuildPattern(
+		memo.OperandAggregation,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandTableGather, memo.EngineTiDBOnly),
+	)
+}
+
+// Match implements Transformation interface.
+func (r *PushAggDownGather) Match(expr *memo.ExprIter) bool {
+	agg := expr.GetExpr().ExprNode.(*plannercore.LogicalAggregation)
+	for _, aggFunc := range agg.AggFuncs {
+		if aggFunc.Mode != aggregation.CompleteMode {
+			return false
+		}
+	}
+	childEngine := expr.Children[0].GetExpr().Children[0].EngineType
+	return plannercore.CheckAggCanPushCop(agg.SCtx(), agg.AggFuncs, agg.GroupByItems, childEngine == memo.EngineTiFlash)
```

Comment:
It does not support tiflash currently, so if the child engine is tiflash, it should not match?

Reply:
Yeah, I just add a check for the childEngine.